openssl/goopenssl.h

```diff
@@ -0,0 +1,34 @@
+// This header file describes the OpenSSL ABI as built for use in Go.
+
+#include <stdlib.h> // size_t
+
+#include "shims.h"
+
+int go_openssl_version_major(void* handle);
+int go_openssl_version_minor(void* handle);
+int go_openssl_thread_setup(void);
+void go_openssl_load_functions(void* handle, int major, int minor);
+
+// Define pointers to all the used OpenSSL functions.
+// Calling C function pointers from Go is currently not supported.
+// It is possible to circumvent this by using a C function wrapper.
+// https://pkg.go.dev/cmd/cgo
+#define DEFINEFUNC(ret, func, args, argscall)      \
+    extern ret (*_g_##func)args;                   \
+    static inline ret go_openssl_##func args       \
```

Comment:
If possible (depending on the C compilers we support), I would suggest using the `typeof` keyword so we don't need to spell out function signatures. That would also prevent any mismatches in function signatures between `goopenssl.h` and OpenSSL headers, as they should be caught at compile time.

[Here](https://github.com/ueno/go/commit/b85fb8abe0298b556a209f10de99eeb0a1239682) is my previous attempt; another example can be found in [libbpf](https://github.com/libbpf/libbpf/blob/10a32130e727719a69e1b2b10815af7c2e245be7/src/usdt.bpf.h#L235).

Reply:
A couple of thoughts:
- CGO internally uses `__typeof__` to generate C bindings, so it can't be used with a C compiler which doesn't support that operand. That's the case of MSVC.
- I like the `typeof` approach, but this PR defines the OpenSSL symbols on a way that OpenSSL are not needed to facilitate OpenSSL portability (e.g., building with 1.1.1 and running with 3.0.1), so I'm not sure if `typeof` would work here.